src/dbg/advancedanalysis.cpp

```diff
@@ -0,0 +1,373 @@
+
+                //add node to the function graph
+                graph.AddNode(node);
+
+                //enqueue branch destinations
+                if(node.brtrue)
+                    queue.push(node.brtrue);
+                if(node.brfalse)
+                    queue.push(node.brfalse);
+
+                break;
+            }
+            if(mCp.InGroup(CS_GRP_CALL))   //call
+            {
+                //TODO: handle no return
+                duint target = mCp.BranchDestination();
+                if(inRange(target) && mEntryPoints.find(target) == mEntryPoints.end())
+                    mCandidateEPs.insert(target);
+            }
+            if(mCp.InGroup(CS_GRP_RET))   //return
+            {
+                node.terminal = true;
+                graph.AddNode(node);
+                break;
+            }
+            node.end += mCp.Size();
+        }
+    }
+    mFunctions.push_back(graph);
+}
+
+
+void AdvancedAnalysis::linearXrefPass()
+{
+    dputs("Starting xref analysis...");
+    auto ticks = GetTickCount();
+
+    for(auto addr = mBase; addr < mBase + mSize;)
+    {
+        if(!mCp.Disassemble(addr, translateAddr(addr)))
+        {
+            addr++;
+            continue;
+        }
+        addr += mCp.Size();
+
+        XREF xref;
+        xref.valid = true;
+        xref.addr = 0;
+        xref.from = mCp.Address();
+        for(auto i = 0; i < mCp.OpCount(); i++)
+        {
+            duint dest = mCp.ResolveOpValue(i, [](x86_reg)->size_t
+            {
+                return 0;
+            });
+            if(inRange(dest))
+            {
+                xref.addr = dest;
+                break;
+            }
+        }
+        if(xref.addr)
+        {
+            if(mCp.InGroup(CS_GRP_CALL))
+                xref.type = XREF_CALL;
+            else if(mCp.InGroup(CS_GRP_JUMP))
+                xref.type = XREF_JMP;
+            else
+                xref.type = XREF_DATA;
+
+            auto found = mXrefs.find(xref.addr);
+            if(found == mXrefs.end())
+            {
+                std::vector<XREF> vec;
+                vec.push_back(xref);
+                mXrefs[xref.addr] = vec;
+            }
+            else
+            {
+                found->second.push_back(xref);
+            }
+        }
+    }
+
+    dprintf("%u xrefs found in %ums!\n", mXrefs.size(), GetTickCount() - ticks);
+}
+
+void AdvancedAnalysis::findInvalidXrefs()
+{
+    for(auto & vec : mXrefs)
+    {
+        duint jmps = 0, calls = 0;
+        duint addr = vec.first;
+        byte desttype = mEncMap[vec.first - mBase];
+        for(auto & xref : vec.second)
+        {
+            byte type = mEncMap[xref.from - mBase];
+            if(desttype == enc_code && type != enc_unknown && type != enc_code)
+                xref.valid = false;
+            else if(desttype == enc_middle)
+                xref.valid = false;
+        }
+    }
+}
+
+bool isFloatInstruction(x86_insn opcode)
```

Comment:
Yes, it's a work in progress, I plan to add simd instruction as well, but it takes a lot of time to look up the manuals.